// es6-transpiler/build/es5/transpiler/spread.js
"use strict";

var assert = require("assert");
var error = require("./../lib/error");
var core = require("./core");


function isSpreadElement(node) {
	return node && node.type === 'SpreadElement';
}

function hasSpreadArgument(node) {
	var elements;
	return node && Array.isArray(elements = (node.elements || node["arguments"])) && elements.some(isSpreadElement);
}

var callIteratorBody =
	"(v,f){" +
		"if(v){" +
			"if(Array.isArray(v))return f?v.slice():v;" +
			"if(typeof v==='object'&&typeof v['iterator']==='function')return Array['from'](v);" +
		"}" +
		"throw new Error(v+' is not iterable')"+
	"};"
;

var plugin = module.exports = {
	reset: function() {

	}

	, setup: function(alter, ast, options) {
		if( !this.__isInit ) {
			this.reset();
			this.__isInit = true;
		}

		this.alter = alter;
		this.options = options;
	}

	, pre: function(node) {
		var type = node.type;
		if( (
				type === "CallExpression"
				|| type === "NewExpression"
				|| type === "ArrayExpression"
			)
			&& hasSpreadArgument(node)
		) {
			if ( type === "CallExpression" ) {
				this.replaceCallExpression(node);
			}
			else if ( type === "NewExpression" ) {
				this.replaceNewExpression(node);
			}
			else if ( type === "ArrayExpression" ) {
				this.replaceArrayExpression(node);
			}
		}
	}

	, replaceCallExpression: function(node) {
		var isMemberExpression = node.callee.type === "MemberExpression";
		var isSimpleMemberExpression = isMemberExpression && node.callee.object.type === "Identifier";
//		const functionNameNode = isMemberExpression ? node.callee.property : node.callee;
		var expressionString;
		var tempVar;
		var valueNode = node["arguments"];

		var expressionBody = this.alter.get(node.callee.range[0], node.callee.range[1]);

		if( isMemberExpression ) {
			if( isSimpleMemberExpression ) {
				expressionString =
					expressionBody
					+ ".apply("
					+ this.alter.get(node.callee.object.range[0], node.callee.object.range[1])
					+ ", "
				;
			}
			else {
				var startsFrom = node.callee.object.range[0]
					, endsFrom = node.range[1]
				;

				tempVar = core.getScopeTempVar(startsFrom, node.$scope);

				expressionString =
					"(" + tempVar + " = "
						+ this.alter.get(startsFrom, node.callee.object.range[1])
						+ ")"
						+ core.PropertyToString(node.callee.property)
						+ ".apply(" + tempVar
						+ ", "
				;

				core.setScopeTempVar(tempVar, endsFrom, node.$scope);
			}
		}
		else {
			if( node.callee.type === "FunctionExpression" ) {
				expressionBody = "(" + expressionBody + ")";
			}
			expressionString =
				expressionBody
				+ ".apply(null, "
			;
		}

		if( valueNode.length === 1 ) {
			var isSequenceExpression = valueNode[0].type === "SequenceExpression";
			var callIteratorFunctionName = core.bubbledVariableDeclaration(node.$scope, "ITER", callIteratorBody, true);
			expressionString += (
				callIteratorFunctionName
				+ "(" + (isSequenceExpression ? "(" : "")
				+ this.alter.get(valueNode[0].range[0] + 3, valueNode[0].range[1])
				+ ")" + (isSequenceExpression ? ")" : "")
			);
		}
		else {
			expressionString += this.__unwrapSpread(node, valueNode);
		}

		this.alter.replace(
			node.range[0]
			, node.range[1]
			, expressionString + ")"
		);
	}

	, replaceNewExpression: function(node) {
		var bindFunctionName = core.bubbledVariableDeclaration(node.$scope, "BIND", "Function.prototype.bind");

		var expressionString =
			"(" + bindFunctionName + ".apply(" + this.alter.get(node.callee.range[0], node.callee.range[1]) + ", "
			+ this.__unwrapSpread(node, node["arguments"], "null")
		;

		this.alter.replace(
			node.callee.range[0]
			, node.range[1]
			, expressionString + ")"
		);

		this.alter.insertAfter(
			node.range[1]
			, ")()"
		);
	}

	, replaceArrayExpression: function(node) {
		// found new line symbols
		var str =
			this.alter._source.substring(node.range[0], node.range[1])//TODO this.alter.getSource(node.range[0], node.range[1])
		;
		var newLines = str.match(/(\n\r)|(\n)|(\r)/g);

		var arrayExpressionStr = this.__unwrapSpread(node) + (newLines ? newLines.join("") : "");
		this.alter.replace(
			node.range[0]
			, node.range[1]
			, arrayExpressionStr
		);
	}

	,
	__unwrapSpread: function(node, elements, firstElementString) {
		elements = elements || node.elements;

		var argsLength = elements.length;
		var spreadIndex = 0;
		var callIteratorFunctionName = core.bubbledVariableDeclaration(node.$scope, "ITER", callIteratorBody, true);

		elements.some(function(arg, index) {
			if( isSpreadElement(arg) ) {
				spreadIndex = index;
				return true;
			}
		});

		firstElementString = firstElementString || "";

		var expressionString
			, nonSpreadStart
			, nonSpreadEnd = null
		;

		if( spreadIndex > 0 ) {
			expressionString =
				"["
				+ (firstElementString ? (firstElementString + ", ") : "")
				+ this.alter.get(elements[0].range[0], elements[spreadIndex - 1].range[1]) + "]"
			;
		}
		else {
			expressionString = "[" + firstElementString + "]";
		}

		expressionString += ".concat(";

		function addNonSpreadStr() {
			var str = "", nodeStart, nodeEnd, isStart = true;

			if( nonSpreadStart ) {
				str += ", [";

				if( nonSpreadEnd === null ) {
					nonSpreadEnd = nonSpreadStart;
				}

				while( (nodeStart = elements[nonSpreadStart]) === null ) {
					str += ",";
					nonSpreadStart++;
				}

				if( nonSpreadStart > nonSpreadEnd ) {
					// next element is Spread Element
				}
				else {
					while( (nodeEnd = elements[nonSpreadEnd]) === null ) {
						str += ",";
						nonSpreadEnd--;
					}

					isStart = true;
					while( nonSpreadStart <= nonSpreadEnd ) {
						str += ((isStart ? "" : ", ") + core.unwrapNode(elements[nonSpreadStart]));
						nonSpreadStart++;
						isStart = false;
					}
				}

				str += "]";

//				expressionString += (", [" + str + (nonSpreadStart !== nonSpreadEnd ? alter.get(nonSpreadStart, nonSpreadEnd) : "") + "]")
				expressionString += str;
			}

			nonSpreadStart = nonSpreadEnd = null;
		}

		var spreadExpressionsCount = 0;
		var currentSpread = 0;

		for(var currentIndex = spreadIndex ; currentIndex < argsLength ; currentIndex++ ) {
			var arg = elements[currentIndex];

			if( arg && arg.type !== "Literal" && arg.type !== "Identifier" ) {
				spreadExpressionsCount++;
			}
		}

		for(var currentIndex$0 = spreadIndex ; currentIndex$0 < argsLength ; currentIndex$0++ ) {
			var arg$0 = elements[currentIndex$0]
				, isSpread = isSpreadElement(arg$0)
				, spreadTypeIsArrayExpression = void 0
			;

			if( isSpread ) {
				addNonSpreadStr();
			}
			else {
				if( nonSpreadStart ) {
					nonSpreadEnd = currentIndex$0;
				}
				else {
					if( arg$0 && arg$0.type === "Literal" ) {
						expressionString += (
							(currentIndex$0 ? ", " : "")
							+ arg$0.raw
						);
					}
					else {
						nonSpreadStart = currentIndex$0;
					}
				}
			}

			if( isSpread ) {
				currentSpread++;
				spreadTypeIsArrayExpression = arg$0.argument.type === "ArrayExpression";
				var isSequenceExpression = arg$0.argument.type === "SequenceExpression";
				var forcedCopyFlag = currentSpread < spreadExpressionsCount;

				//console.log(spreadIndex, argsLength)

				expressionString += (
					(currentIndex$0 !== spreadIndex ? ", " : "")
					+ (spreadTypeIsArrayExpression ? "" : callIteratorFunctionName + "(")
						+ (isSequenceExpression ? "(" : "")
					+ this.alter.get(arg$0.argument.range[0], arg$0.argument.range[1])
						+ (isSequenceExpression ? ")" : "")
					+ (spreadTypeIsArrayExpression ? "" : (forcedCopyFlag ? ", true" : "") + ")")
				);
			}
		}

		if( nonSpreadStart ) {
			addNonSpreadStr();
		}

		expressionString += ")";

		return expressionString;
	}
};

for(var i in plugin) if( plugin.hasOwnProperty(i) && typeof plugin[i] === "function" ) {
	plugin[i] = plugin[i].bind(plugin);
}
